cpp/src/arrow/python/numpy_to_arrow.cc

```diff
@@ -554,12 +554,22 @@ Status StaticCastBuffer(const Buffer& input, const int64_t length, MemoryPool* p
   return Status::OK();
 }
 
-template <typename T, typename T2>
-void CopyStrided(T* input_data, int64_t length, int64_t stride, T2* output_data) {
+template <typename T>
+void CopyStridedBytewise(int8_t* input_data, int64_t length, int64_t stride,
+                         T* output_data) {
+  // Passing input_data as non-const is a concession to PyObject*
+  for (int64_t i = 0; i < length; ++i) {
+    memcpy(output_data + i, input_data, sizeof(T));
+    input_data += stride;
```

Comment:
Is `stride >= sizeof(T)` guaranteed?

Reply:
In normal use, it probably is, but you can create weird arrays using `stride_tricks`: https://docs.scipy.org/doc/numpy/reference/generated/numpy.lib.stride_tricks.as_strided.html

That said, I don't think it makes a problem here, does it?

Reply:
No, but maybe add a DCHECK?

Reply:
Well, the current code actually works as intended even with such arrays:
```
>>> base = np.arange(8, dtype=np.int8).view(np.int32)
>>> arr = np.lib.stride_tricks.as_strided(base, shape=(5,), strides=(1,))
>>> arr
array([ 50462976,  67305985,  84148994, 100992003, 117835012], dtype=int32)
>>> arr.tobytes()
b'\x00\x01\x02\x03\x01\x02\x03\x04\x02\x03\x04\x05\x03\x04\x05\x06\x04\x05\x06\x07'
>>> pa.array(arr, type=pa.int32())
<pyarrow.lib.Int32Array object at 0x7ff7e9094098>
[
  50462976,
  67305985,
  84148994,
  100992003,
  117835012
]
```

Reply:
Fair enough. LGTM